lib/Sema/CSBindings.cpp

```diff
@@ -1090,11 +1090,26 @@ bool TypeVariableBinding::attempt(ConstraintSystem &cs) const {
       ConstraintFix *fix = nullptr;
       if (auto *GP = TypeVar->getImpl().getGenericParameter()) {
         auto path = dstLocator->getPath();
-        // Drop `generic parameter` locator element so that all missing
-        // generic parameters related to the same path can be coalesced later.
-        fix = DefaultGenericArgument::create(
-            cs, GP,
-            cs.getConstraintLocator(dstLocator->getAnchor(), path.drop_back()));
+
+        auto result = cs.getConstraintGraph().lookupNode(TypeVar);
+        auto equivalence = result.first.getEquivalenceClass();
+        // Let's try to emit a specialized fix if the generic argument is
+        // connected with a key path root type variable.
+        auto kpRootEq = llvm::find_if(equivalence, [](TypeVariableType *eq) {
+          auto loc = eq->getImpl().getLocator();
+          return loc && loc->isKeyPathRoot();
+        });
+        if (kpRootEq != equivalence.end()) {
+          fix = SpecifyKeyPathRootType::create(
+              cs, (*kpRootEq)->getImpl().getLocator());
```

Comment:
> Type variable representing key path root doesn't even show up in "considered" bindings while seems weird to me.

Humm... I'd go that way because I thought that for example `$T1`(root type) and `$T0` (generic param), `$T1` wouldn't be bound directly because as my understanding `$T0` is a representative for `$T1`.
Another approach I thought was to record the fix directly in `simplifyKeypathConstraint` if root is a hole... maybe could be an option...

I'll take a look later today, thanks @xedin :)

Reply:
Sure!

> Another approach I thought was to record the fix directly in simplifyKeypathConstraint if root is a hole

If you do this, you'd have to bind key path root to a hole there as well.

Reply:
Hey @xedin :)
> If you do this, you'd have to bind key path root to a hole there as well.

Do you mean actually adding a binding constraint(creating a Bind constraint or assigning the fixed type)? 

Also, I'm just back working on this and notice that actually at the point we got into `simplifyKeyPathConstraint` with the rootFixed type as a hole, the generic `DefaultGenericArgument` fix is already recorded... So, we could record the fix in there, but I don't know how to not record the generic fix at this point...

Reply:
> Do you mean actually adding a binding constraint(creating a Bind constraint or assigning the fixed type)?

Yes, I mean actually bind it to `UnresolvedType` :)

> generic DefaultGenericArgument fix is already recorded.

That's what I was referring in relation to CSBindings - it seems like solver only attempts to bind generic parameter to a hole in that example, instead we should discover that it's possible to bind key path root type to a hole before generic parameter.

Reply:
> That's what I was referring in relation to CSBindings - it seems like solver only attempts to bind generic parameter to a hole in that example

Yes, I'm curious about that ... but the solver should attempt to bind the key path root type variable to hole in this case? I mean once it bound its representative(generic argument) type variable it. The mental model I had of this(let me know if it is correct) is that once its representative has been bound to a fixed type it kinda doesn't need to attempt bound it. Or should it attempt to bind it anyways? 

In this example, we have
$T0 -> (Key Path type)
$T1 -> (Generic Argument)
$T2 -> (Key Path root)

The initial binding is $T0 := (($T1) -> String) -> ())
After open the generic type $T1, it just merges equivalence classes for $T1 -> $T2 and tries to bind the initial $T1 

I think I need to get a better understanding of how the binding is gathered... maybe if the initial binding was ` $T0 := (($T2) -> String) -> ())` we could have key path root being bounded before, but I don't know exactly how it would play out with the generic argument

Reply:
Better thinking about it I think the solver wouldn't attempt binding key path type variable in this case since it is a non-representative...

Reply:
Hey @xedin :) 
I just made some changes, to ensure diagnosing the inability of the solver to infer the root even if the type variable is non-representative. This way we end up producing both diagnostics, which I think is ok, but anyways let me know what you think

Reply:
How about we take a different approach here and check in diagnostics for generic parameter whether it's a representative for a key path root and if so diagnose as a failure to infer key path root?

Reply:
Sure! Seems a good approach to me, I'll adjust it :)